source/simulator.cpp

```diff
@@ -0,0 +1,80 @@
+#include "simulator.hpp"
+
+#include "link.hpp"
+#include "receiver.hpp"
+#include "scheduler.hpp"
+#include "sender.hpp"
+#include "switch.hpp"
+
+namespace sim {
+
+Simulator::Simulator() : m_scheduler(Scheduler::get_instance()) {}
+
+Simulator::~Simulator() {
+    for (auto& p : m_graph) {
+        delete p.second;
+    }
+}
+
+void Simulator::add_device(std::string a_name, DeviceType a_type) {
+    if (m_graph.find(a_name) != m_graph.end()) {
+        return;
```

Comment:
I think we might want some log message here
```suggestion
        logger->warn("Device with such name already exists.");
        return;
```

Reply:
Now this method returns `Device*` and in case of name collision a `nullptr` will be returned.